TrackController/src/trackctrl/model/TrackController.java

```diff
@@ -92,54 +108,83 @@ public void setZone(HashMap<Integer, Block> blocks) {
 
       //Create string list upon initialization
       for (Integer i : blocks.keySet()) {
-        blockList.add(i.toString());
+        blockList.add(i);
       }
     }
   }
 
   @Override
-  public ArrayList<String> getZone() {
+  public ArrayList<Integer> getZone() {
+
+    for (Map.Entry<Integer, Block> b : myZone.entrySet()) {
+      blockList.add(b.getKey());
+    }
     return blockList;
   }
 
-  //TODO
+  @Override
+  public int getBlockCount() {
+    return blockList.size();
+  }
+
+  @Override
+  public boolean getOccupancy(int id) {
+    return myLine.getBlock(id).isOccupied();
+  }
+
   @Override
   public boolean addBlock(Block newBlock) {
-    return false;
+
+    myZone.put(newBlock.getNumber(), newBlock);
+    return myZone.containsValue(newBlock);
   }
 
   @Override
   public boolean hasBlock(int id) {
     return myZone.containsKey(id);
   }
 
-  //TODO
   @Override
   public boolean closeBlock(int id) {
-    return false;
+    myLine.getBlock(id).setBrokenRailStatus(true);
+    return myLine.getBlock(id).getBrokenRailStatus() == true;
   }
 
-  //TODO
   @Override
   public boolean repairBlock(int id) {
-    return false;
+    myLine.getBlock(id).setBrokenRailStatus(false);
+    return myLine.getBlock(id).getBrokenRailStatus() == false;
   }
 
-  //TODO
   @Override
   public boolean toggleSwitch(int id) {
+    Block toggle = myLine.getBlock(id);
+    if (toggle.isSwitch()) {
+      Switch toggleSwitch = (Switch) toggle;
+      boolean before = toggleSwitch.getSwitchState();
+      toggleSwitch.toggle();
+      boolean after = toggleSwitch.getSwitchState();
+      return before & after;
+    }
```

Comment:
This should be implemented and called from the Track. The stub out is in already there